### PR TITLE
7903193: [jtreg] build and test failures using JDK 18

### PR DESCRIPTION
**Pull request by jaikiran**
Can I please get a review of this test only changes, which proposes to address the current failure in jtreg self tests when Java 18 or higher is used to build and test jtreg?

As noted in the following issues:
https://bugs.openjdk.org/browse/CODETOOLS-7903193
https://bugs.openjdk.org/browse/CODETOOLS-7903646
https://bugs.openjdk.org/browse/CODETOOLS-7903645

these self tests in jtreg which rely on SecurityManager, no longer pass when used with Java 18 or higher, since starting Java 18 the setting of SecurityManager throws an UnsupportedOperationException.

Changes in this PR, include updates to test files which check for the Java version being used to run these tests and then decide whether or not to include some specific tests that only pass when a SecurityManager is set.

I've run these changes locally (on macos M1) and on a linux setup, both with Java 17  and Java 21. The tests all pass on these versions.

I've also run this on a headless system to make sure the `ReportOnlyTest.gmk` does indeed properly check the correct values on a headless system (both Java 17 and 21). I think this change should address the issue that Ludvig @LudwikJaniuk had run into.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issues
 * [CODETOOLS-7903193](https://bugs.openjdk.org/browse/CODETOOLS-7903193): [jtreg] build and test failures using JDK 18 (**Bug** - P3)
 * [CODETOOLS-7903645](https://bugs.openjdk.org/browse/CODETOOLS-7903645): tests that set SecurityManager fail due to UnsupportedOperationException in Java 21 (**Bug** - P4)
 * [CODETOOLS-7903646](https://bugs.openjdk.org/browse/CODETOOLS-7903646): test/BasicAgent.vm fails when run with Java 21 (**Bug** - P3)

### Reviewers
 * [Christian Stein](https://openjdk.org/census#cstein) (@sormuras - Committer)
 * [Jan Lahoda](https://openjdk.org/census#jlahoda) (@lahodaj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jtreg.git pull/190/head:pull/190` \
`$ git checkout pull/190`

Update a local copy of the PR: \
`$ git checkout pull/190` \
`$ git pull https://git.openjdk.org/jtreg.git pull/190/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 190`

View PR using the GUI difftool: \
`$ git pr show -t 190`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jtreg/pull/190.diff">https://git.openjdk.org/jtreg/pull/190.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jtreg/pull/190#issuecomment-1987040686)

### CODE REVIEW

File: make/build.sh [hunk 1]
```diff
@@ -784,6 +785,7 @@ make ASMTOOLS_JAR="${ASMTOOLS_JAR}"                           \\
      JUNIT_NOTICES="${JUNIT_NOTICES}"                         \\
      TESTNG_JARS="${TESTNG_JARS}"                             \\
      TESTNG_NOTICES="${TESTNG_NOTICES}"                       \\
+     JAVA_SPECIFICATION_VERSION="${JAVA_SPECIFICATION_VERSION}" \\
```
Comment (jonathan-gibbons) [inline make/build.sh:788]: Note the list of variables is sorted alphabetically.

File: test/basic/ReportOnlyTest.gmk [hunk 2]
```diff
@@ -40,6 +35,7 @@ $(BUILDTESTDIR)/ReportOnlyTest.ok: $(BUILDTESTDIR)/Basic.othervm.ok \
 		   $(JTREG_IMAGEDIR)/bin/jtreg
 	JAVA_HOME=$(JDKHOME) CLASSPATH=$(ABSCLASSDIR) \
 	    $(JTREG_IMAGEDIR)/bin/jtreg $(JTREG_OPTS) \
+	    -k:!needDisplay \
```
Comment (jonathan-gibbons) [inline test/basic/ReportOnlyTest.gmk:38]: This seems unrelated to the security manager changes.

**Review by jonathan-gibbons — COMMENTED** [review]
It would be better to come up with a better way for `jtreg` to deal with the lack of a security manager.

That being, said, we obviously _are_ using `jtreg` to run on recent versions of JDK, and this PR is just about fixing the tests, not about changing `jtreg` functionality itself, at this time.

File: test/basic/ReportOnlyTest.gmk (hunk 2 above)
Comment (jaikiran) [reply test/basic/ReportOnlyTest.gmk:38]: Sorry Jon, I forgot to reply to this comment. I'll refresh this PR against latest master branch and refresh my memory about this change and respond shortly.

**Comment by jaikiran** [comment]
I've now updated the PR to only contain the Java 18+ changes. I have verified the latest changes against Java versions less than 18 as well as greater than 18, both when headless is true and when headless is false. All self tests continue to pass and jtreg builds successfully in all these combinations.

File: test/basic/ReportOnlyTest.gmk (hunk 2 above)
Comment (jaikiran) [reply test/basic/ReportOnlyTest.gmk]: With the other PR https://github.com/openjdk/jtreg/pull/223 now integrated, this part of the change in this PR is no longer needed. So I've removed it in the latest updates to this PR.

**Comment by sormuras** [comment]
Heads-up for after https://github.com/openjdk/jtreg/pull/217 has been integrated, the basic test number counters need an update, again.

**Comment by jaikiran** [comment]
I've now updated the PR after merging against latest master. I have verified that with these changes, all self tests continue to pass with Java versions less than 18 as well as greater than 18, both when headless is true and when headless is false.

**Comment by sormuras** [comment]
@jaikiran Can you please replace the `17` with a `21` in https://github.com/jaikiran/jtreg/blob/60809c4da0f2ef18540df4328113f898d032497b/.github/workflows/test.yml#L24

That should make a) the test workflow download an Oracle JDK and b) also run the build (compilation and tests) on a JDK 18+.

**Comment by jaikiran** [comment]
> @jaikiran Can you please replace the 17 with a 21 in https://github.com/jaikiran/jtreg/blob/60809c4da0f2ef18540df4328113f898d032497b/.github/workflows/test.yml#L24
>
> That should make a) the test workflow download an Oracle JDK and b) also run the build (compilation and tests) on a JDK 18+.

I wasn't sure if we should do that change as part of this PR, so I created a separate branch which includes the changes in this PR plus the change to the JDK version in the github workflow file. That branch's job is currently in progress here https://github.com/jaikiran/jtreg/actions/runs/11385998691/job/31677151971

**Comment by jaikiran** [comment]
If that build succeeds and if you think that we should include that change in this PR, let me know and I'll update this PR to include that change.

**Comment by sormuras** [comment]
> If that build succeeds and if you think that we should include that change in this PR, let me know and I'll update this PR to include that change.

Sounds good to me.

**Review by sormuras — APPROVED** [review]
All green with JDK 21

**Review by lahodaj — APPROVED** [review]
Seems fine to me.

I'll note that you can enable the `setSecurityManager` with `-Djava.security.manager=allow` on the command line, for JDK 18+, until https://github.com/openjdk/jdk/pull/21498 is integrated. But there presumably is no strong reason to do so for the tests.

**Comment by jaikiran** [comment]
Thank you all for reviewing this PR. 

> I'll note that you can enable the setSecurityManager with -Djava.security.manager=allow on the command line, for JDK 18+, until https://github.com/openjdk/jdk/pull/21498 is integrated. But there presumably is no strong reason to do so for the tests.

Right, supporting the ability to run these self tests on Java 18+ with securtiy manager enabled would require additional work which can be done but didn't feel worth the efforts.
